bindings/rust/evmc-client/src/host.rs

```diff
@@ -0,0 +1,237 @@
+    context: *mut ffi::evmc_host_context,
+    address: *const ffi::evmc_address,
+) -> ffi::evmc_bytes32 {
+    return ffi::evmc_bytes32 {
+        bytes: (*(context as *mut ExtendedContext))
+            .hctx
+            .get_code_hash(&(*address).bytes),
+    };
+}
+
+unsafe extern "C" fn copy_code(
+    context: *mut ffi::evmc_host_context,
+    address: *const ffi::evmc_address,
+    code_offset: usize,
+    buffer_data: *mut u8,
+    buffer_size: usize,
+) -> usize {
+    return (*(context as *mut ExtendedContext)).hctx.copy_code(
+        &(*address).bytes,
+        &code_offset,
+        &buffer_data,
+        &buffer_size,
+    );
+}
+
+unsafe extern "C" fn selfdestruct(
+    context: *mut ffi::evmc_host_context,
+    address: *const ffi::evmc_address,
+    beneficiary: *const ffi::evmc_address,
+) {
+    (*(context as *mut ExtendedContext))
+        .hctx
+        .selfdestruct(&(*address).bytes, &(*beneficiary).bytes)
+}
+
+unsafe extern "C" fn get_tx_context(context: *mut ffi::evmc_host_context) -> ffi::evmc_tx_context {
+    let (gas_price, origin, coinbase, number, timestamp, gas_limit, difficulty, chain_id) =
+        (*(context as *mut ExtendedContext)).hctx.get_tx_context();
+    return ffi::evmc_tx_context {
+        tx_gas_price: evmc_sys::evmc_bytes32 { bytes: gas_price },
+        tx_origin: evmc_sys::evmc_address { bytes: origin },
+        block_coinbase: evmc_sys::evmc_address { bytes: coinbase },
+        block_number: number,
+        block_timestamp: timestamp,
+        block_gas_limit: gas_limit,
+        block_difficulty: evmc_sys::evmc_bytes32 { bytes: difficulty },
+        chain_id: evmc_sys::evmc_bytes32 { bytes: chain_id },
+    };
+}
+
+unsafe extern "C" fn get_block_hash(
+    context: *mut ffi::evmc_host_context,
+    number: i64,
+) -> ffi::evmc_bytes32 {
+    return ffi::evmc_bytes32 {
+        bytes: (*(context as *mut ExtendedContext))
+            .hctx
+            .get_block_hash(number),
+    };
+}
+
+unsafe extern "C" fn emit_log(
+    context: *mut ffi::evmc_host_context,
+    address: *const ffi::evmc_address,
+    data: *const u8,
+    data_size: usize,
+    topics: *const ffi::evmc_bytes32,
+    topics_count: usize,
+) {
+    let ts = &std::slice::from_raw_parts(topics, topics_count)
+        .iter()
+        .map(|topic| topic.bytes)
+        .collect::<Vec<_>>();
+    (*(context as *mut ExtendedContext)).hctx.emit_log(
+        &(*address).bytes,
+        &ts,
+        &std::slice::from_raw_parts(data, data_size),
+    );
+}
+
+unsafe extern "C" fn release(result: *const ffi::evmc_result) {
+    drop(std::slice::from_raw_parts(
+        (*result).output_data,
+        (*result).output_size,
+    ));
+}
+
+pub unsafe extern "C" fn call(
+    context: *mut ffi::evmc_host_context,
+    msg: *const ffi::evmc_message,
+) -> ffi::evmc_result {
+    let msg = *msg;
+    let (output, gas_left, create_address, status_code) =
+        (*(context as *mut ExtendedContext)).hctx.call(
```

Comment:
This entire function could be reduced to `hctx.call().into()` is using `ExecutionResult` from `evmc-vm/lib.rs` because that implements the translation and memory handling already.